### ISSUE TITLE
bug in 2.0.8: failing view tests and a compile warning

### ISSUE THREAD

**Issue by ckruse** [appsignal/appsignal-elixir-phoenix#20]
After updating to 2.0.8 I get this compiler warning for some of my views:

~~~
warning: this clause for render (overridable 1)/2 cannot match because a previous clause at line 1 always matches
~~~

Additionally all `render` calls in my view tests just return `nil`, e.g. this test:

~~~elixir
  test "renders 500.json" do
    assert render(TermitoolWeb.ErrorView, "500.json", []) == %{errors: %{detail: "Internal Server Error"}}
  end
~~~

now just returns `nil`:

~~~
  1) test renders 500.json (TermitoolWeb.ErrorViewTest)
     test/termitool_web/views/error_view_test.exs:11
     Assertion with == failed
     code:  assert render(TermitoolWeb.ErrorView, "500.json", []) == %{errors: %{detail: "Internal Server Error"}}
     left:  nil
     right: %{errors: %{detail: "Internal Server Error"}}
     stacktrace:
       test/termitool_web/views/error_view_test.exs:12: (test)
~~~

After downgrading to 2.0.7 everything works again.

**Comment by szymon-jez** [comment]
I can confirm that upgrading to 2.0.8 breaks tests. I have seen it in 2 applications.

**Comment by shairyar** [comment]
Thanks for reporting the issue, we will take a look.

**Comment by jeffkreeftmeijer** [comment]
Thanks for reporting this, both. This is a regression inn 2.0.8, and we’re reviewing a fix for this in https://github.com/appsignal/appsignal-elixir-phoenix/pull/21. 

We’re aiming to release 2.0.9 with this patch soon. In the meantime, please downgrade to 2.0.7.

**Comment by jeffkreeftmeijer** [comment]
Hey all, I’ve just published version 2.0.9 of the Phoenix integration, which is a patch to resolve the issues introduced in 2.0.8. Could you upgrade and let me know if this resolves the issue for you?

Thanks again!

**Comment by szymon-jez** [comment]
@jeffkreeftmeijer Thanks for the quick reply.

I have checked out the new version and now regrettably my test suite prints out tens of such warnings:

```
warning: this clause for render/2 cannot match because a previous clause at line 1 always matches
  web/views/terms_of_service.ex:1
```

**Comment by jeffkreeftmeijer** [comment]
I’ll take another look to see what’s causing that after the change in 2.0.8. Do you still get `nil` responses from your error views with 2.0.9?

**Comment by szymon-jez** [comment]
> Do you still get nil responses from your error views with 2.0.9?

CI does not even get there as if stops at compilation on those warnings. I will check locally and report back.

**Comment by jeffkreeftmeijer** [comment]
Thanks! I’ve pushed a fix for the warnings, which is working for me in my local setup: https://github.com/appsignal/appsignal-elixir-phoenix/pull/22. If time permits, could you try the `view-warnings` for appsignal_phoenix to see if that still produces those warnings? We’ll do one more round of testing before pushing this out.

**Comment by szymon-jez** [comment]
So for `2.0.9` I get:

```
** (FunctionClauseError) no function clause matching in IO.chardata_to_string/1

     The following arguments were given to IO.chardata_to_string/1:
     
         # 1
         Clubbase.Views.Layout.Club
     
     Attempted function clauses (showing 2 out of 2):
     
         def chardata_to_string(string) when is_binary(string)
         def chardata_to_string(list) when is_list(list)
     
     code: conn = get(conn, WebRoutes.club_member_signups_path(conn, :index, club))
     stacktrace:
       (elixir 1.11.4) lib/io.ex:589: IO.chardata_to_string/1
       (elixir 1.11.4) lib/path.ex:530: Path.do_join/3
       (elixir 1.11.4) lib/path.ex:522: Path.join/2
```

but on `branch: "view-warnings"` don't have the above reported compiler warnings and also all tests pass.

**Comment by jeffkreeftmeijer** [comment]
That’s good to hear, thanks for helping us test this. I’ve just released version 2.0.10, which includes the patch in the `view-warnings` branch. Please let me know if you run into any issues with that!

**Comment by ckruse** [comment]
Can confirm, works now again. Thanks for fixing.

**Comment by begedin** [comment]
Still seems to be happening in 2.0.11

running  `mix deps.clean` sometimes helps at random, but not every time and it randomly starts happening after a while. It might be compilation related.

Also, when it does happen, and I run dialyzer, I get warnings along the lines of 

```
lib/{view_module_path}:pattern_match_cov
The pattern
_, _

can never match, because previous clauses completely cover the type
_, _.

________________________________________________________________________________
lib/{view_module_path}:pattern_match_cov
The pattern
{template name as charlist}, %{assigns_map}

can never match, because previous clauses completely cover the type
_, _.
```

I get one error for every render clause I have in my view module. 
It only seems to happen for one view module to.

```
elixir -v
Erlang/OTP 23 [erts-11.2.2.3] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Elixir 1.11.4 (compiled with Erlang/OTP 23)
```

**Comment by tombruijn** [comment]
Hi @begedin, I am unable to reproduce it in a small test app. You say it happens randomly. Does it keep happening consistently after it fails once?

**Comment by begedin** [comment]
@tombruijn I'm afraid it wasn't a consistent reproduction after it happens once.

That being said, I haven't seen it for the past two days, so it's also possible it got resolved by 2.0.12? We have been upgrading as soon as a new version was out.

Maybe it was also some weird caching issue, or interaction with another dependency? We try to keep our stuff up to date with dependabot, so several things got upgraded yesterday.

**Comment by tombruijn** [comment]
It should have been solved in the latest release yes. It may be a caching issue. It doesn't hurt to throw away the `_build` and `deps` directories and any dialyzer cache if it does happen again. It's what I do whenever I worry there's a caching issue 😅
